### ISSUE TITLE
A few comments about stack variables and byte ordering...

### ISSUE THREAD

**Issue by gschmidt958** [RobTillaart/FastShiftIn#15]
Hi, and sorry to bother you once again, but after playing around some more with my project, I have a couple comments:

1. I noticed that rv is an automatic (stack) variable.  For the readxx() calls it is not explicitely initialized to zero and my understanding of the C++ standard is that it does not guarantee that stack variables are initialized to zero.  That won't be a problem for reading a single byte, but for reading multiple bytes, the additional bytes are added into rv, so it seems to me like that could be a problem.

2. When reading multiple bytes, it appears that the byte order is determined by the LSB/MSB first setting.  That's not unreasonable, and in my situation, I wired the 74HCx165 intuitively such that A is bit 0, and H is bit 7 which seemed natural, but actually H is shifted out first, so it really makes H the LSB which is a bit counter-intuitive.  However, my byte order is Chip0 low, Chip1 high which arguably seems natural in the general case because Chip0 data always appears first.  However, I think the byte ordering treats a cascade of registers conceptually as a single large shift register with the LSB/MSB first setting applied to the entire "large" register which is why I think it was done that way.  So in my case, I was able to simply work around the problem by calling read() multiple times and placing the bytes in least significant byte order even though the bit order is MSB for the reason stated above.  Not a big problem, but just something to muse over.  It might be nice for the bit order and byte order to be specified independently, but I'm sure that complicates the library code so I'm not even suggesting that.  I just thought my situation was interesting and it made me realize that bit 0 really corresponds to the "H" pin, not the "A" pin.  Perhaps a bit of a caveat.  Moral: Always read the datasheet, as things are not always as they seem!

Thanks!

**Comment by RobTillaart** [comment]
Thanks for this feedback, appreciated 

1. Agree, better to make initializations explicit.
2. Have to verify the details, lib might need additional documentation about byte order. 

Bitwise the library calls should behave like the reference **shiftIn()**. I will verify / compare the code once again.

**Comment by RobTillaart** [comment]
At lowest level the **rv** variable is initialized 

At the "higher" level functions it is not needed 
```cpp
uint16_t FastShiftIn::read16()
{
  uint16_t rv;
  if (_bitOrder == LSBFIRST)
  {
    rv = readLSBFIRST();  <<<  rv is overwritten so initialization not needed.
    rv += uint16_t(readLSBFIRST()) << 8;
    return rv;
  }
  rv = readMSBFIRST();
  rv <<= 8;
  rv += readMSBFIRST();
  return rv;
}
```

possibly better to rewrite them to

```cpp
uint16_t FastShiftIn::read16()
{
  if (_bitOrder == LSBFIRST)
  {
    uint16_t rv = readLSBFIRST();
    rv += uint16_t(readLSBFIRST()) << 8;
    return rv;
  }
  uint16_t rv = readMSBFIRST();
  rv <<= 8;
  rv += readMSBFIRST();
  return rv;
}

**Comment by gschmidt958** [comment]
Yes, the initial assignment initializes the entire value, so not an issue.  It was late here, when I wrote that.  I think your bit ordering is working fine.  If anything, it might be worth documenting that the LSB for the 74HCx595 is actually "G", not "A".  FWIW, that is what tripped me up on the byte ordering. (I saw a similar post in the Arduino forum where someone encountered that problem, but nothing to do with your library, just a general concern for the 74HCx165).
